File: Custom Python Simulation/PID.py
class PID():

    def __init__(self, kp, ki, kd, setpoint, previousError):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.setpoint = setpoint
        self.integralError = 0
        self.previousError = previousError

    def compute(self, pos, dt):
        error = self.setpoint - pos #compute the error
        derivativeError = (error - self.previousError) / dt 
        self.integralError += error * dt #error build up over time
        output = self.kp*error + self.ki*self.integralError + self.kd*derivativeError 
        self.previousError = error
        #print('%.2f' % dt + " " + '%.2f' % error + " " + '%.2f' % self.integralError + " " + '%.2f' % derivativeError + " " + '%.2f' % output + " " + '%.2f' % self.previousError)
        return output

File: Arduino Code/ThrustVectoring/PID.cpp
#include "PID.h"

void PID::Init(float _Kp, float _Ki, float _Kd) {
  Kp = _Kp;
  Ki = _Ki;
  Kd = _Kd;
  Serial.println(String(Kp) + "\t" + String(Ki) + "\t" + String(Kd));
}

float PID::ComputeCorrection(float error, unsigned long loopTime) {
  if (loopTime == 0) {
    return 0;
  }
  float deltaTime = loopTime / 1000.0;
  integrator += error * deltaTime;
  float derivative = (error - previousError) / deltaTime;
  float output = (Kp * error) + (Ki * integrator) + (Kd * derivative);

  previousError = error;

  Serial.println(String(rocketOrientation) + "\t" + String(error) + "\t" + String(deltaTime) + "\t" + String(integrator) + "\t" + String(derivative) + "\t" + String(output));
  return output;
}

void PID::setSetpoint(float _setpoint) {
  setpoint = _setpoint;
}